Concatting multiple linear layers with same input Tensor (different weight/bias)

Gamrix opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #63198

Summary:
Linear layers using the same input tensor can be concatted together
as long as the weights and biases are compatible.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D31240642](https://our.internmc.facebook.com/intern/diff/D31240642)

Gamrix committed:
Working on concatting multiple linear layers together

Summary:
Linear layers using the same input tensor can be concatted together
as long as the weights and biases are compatible.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Gamrix committed:
Update on "Working on concatting multiple linear layers together"


Summary:
Linear layers using the same input tensor can be concatted together
as long as the weights and biases are compatible.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Gamrix committed:
Update on "Working on concatting multiple linear layers together"


Summary:
Linear layers using the same input tensor can be concatted together
as long as the weights and biases are compatible.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Gamrix committed:
Update on "Working on concatting multiple linear layers together"


Summary:
Linear layers using the same input tensor can be concatted together
as long as the weights and biases are compatible.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Gamrix committed:
Update on "Working on concatting multiple linear layers together"


Summary:
Linear layers using the same input tensor can be concatted together
as long as the weights and biases are compatible.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/frozen_concat_linear.cpp:
```diff
@@ -0,0 +1,186 @@
+
+          // For now we will just keep it simple and require matching types
+          // Type promotion mgiht cause performance to actually decrease.
+          if (weight_type != candidate_weight_type || bias_type != candidate_bias_type) {          
```
nice! we should also be checking be checking device here

eellison commented on torch/csrc/jit/passes/frozen_concat_linear.cpp:
```diff
@@ -0,0 +1,186 @@
+    // backwards.
+    std::unordered_map<Value*, std::list<Node*>> tensorUsers;
+
+    for (Node* n : b->nodes()) {
```
Can you factor out the `collectConstantLinearLayers` logic and the `combineLinearLayers` logic into two separate functions?

eellison commented on torch/csrc/jit/passes/frozen_concat_linear.cpp:
```diff
@@ -0,0 +1,186 @@
+        continue;
+      }
+      
+      // Punting on support for non-constant params for now.
```
`for now` - isnt this somewhat permanent?

eellison replied:
It is somewhat fundamental in that if we don't have constant parameters we would have to concat at runtime, which has a performance penalty

eellison commented on torch/csrc/jit/passes/frozen_concat_linear.cpp:
```diff
@@ -0,0 +1,186 @@
+        }
+
+        // Now actually merge the candidate nodes
+        graph_->setInsertPoint(base_node);
```
this API isnt met to be used because it modifies global state, use WithInsertPoint instead

Gamrix replied:
I didn't see a good way of changing the insert point (because I set this twice) without creating new scopes. 
I could heap allocate and delete a WithInsertPoint object, but that feels weird. 
Putting it in an anonymous scope makes it hard to use the other variables I make.

eellison commented on torch/csrc/jit/passes/frozen_concat_linear.cpp:
```diff
@@ -0,0 +1,186 @@
+      while (!uses.empty()) {
+        auto it_first = uses.begin();
+        Node* base_node = *it_first;
+        uses.erase(it_first);
```
why are we erasing this ?

eellison commented on torch/csrc/jit/passes/frozen_concat_linear.cpp:
```diff
@@ -0,0 +1,186 @@
+        graph_modified = true;
+
+        for (auto n : matching_nodes) {
+          if (n == base_node) {
```
would it be simpler to have `matching_nodes` not contain `base_node`?

Gamrix replied:
No, after that point, all of the uses of that list treat `base_node` as just one part of `compatible_layers` (new name for `matching_nodes`.

eellison commented on torch/csrc/jit/passes/frozen_concat_linear.cpp:
```diff
@@ -0,0 +1,186 @@
+
+        auto weight_type = constant_as<Tensor>(base_node->namedInput("weight"))->dtype();
+        auto bias_type = constant_as<Tensor>(base_node->namedInput("bias"))->dtype();
+        std::vector<Node*> matching_nodes;
```
nit: I think `matching_nodes` is a little bit misleading, since it doesn't make a lot of sense to say a node is matched to itself. maybe `compatible_linear_layers` ?

eellison commented on torch/csrc/jit/passes/frozen_concat_linear.cpp:
```diff
@@ -0,0 +1,186 @@
+        auto weight_list = c10::fmap(
+            matching_nodes, [](Node* n) { return n->namedInput("weight"); });
+        auto weight_with_dim(weight_list);
+        weight_with_dim.push_back(zero);
```
im not sure what we're doing here... getting all weights then why are we adding zero to it?

eellison commented on torch/csrc/jit/passes/frozen_concat_linear.cpp:
```diff
@@ -0,0 +1,186 @@
+            matching_nodes, [](Node* n) { return n->namedInput("bias"); });
+        bias_with_dim.push_back(zero);
+
+        auto weight_cat = graph_->create(prim::VarConcat, weight_with_dim);
```
nit: I would maybe just use concat for now, VarConcat not rly globally used..

eellison replied:
We want to be running concat now, not at runtime

eellison commented on torch/csrc/jit/passes/frozen_concat_linear.cpp:
```diff
@@ -0,0 +1,186 @@
+        linear_node->insertBefore(base_node);
+
+        // Update the outputs of the nodes
+        graph_->setInsertPoint(linear_node);
```
bump to not use `setInsertPoint` here

eellison commented on torch/csrc/jit/passes/frozen_concat_linear.cpp:
```diff
@@ -0,0 +1,186 @@
+        Value* neg1 = graph_->insertConstant(-1);
+        Value* one = graph_->insertConstant(1);
+
+        long long cur_loc = 0;
```
i dont think i've seen `long long` used in pytorch code base at all.. `size_t` ?

Gamrix replied:
is `size_t` or `int64_t` a better choice here. Also looks like due to `long long` not being a standard size, I am getting compiler errors on some targets.

eellison replied:
either is fine probably. size_t i guess since this is always positive

eellison commented on torch/csrc/jit/passes/frozen_concat_linear.cpp:
```diff
@@ -0,0 +1,186 @@
+        }
+        graph_modified = true;
+
+        for (auto n : matching_nodes) {
```
Can we also factor this out into a function?

eellison commented on torch/csrc/jit/passes/frozen_concat_linear.cpp:
```diff
@@ -0,0 +1,186 @@
+        long long cur_loc = 0;
+        Value* cur_val = zero;
+
+        for (auto& orig_node : matching_nodes) {
```
can you add some comments here for the logic  of what we're doing?

eellison commented on torch/csrc/jit/passes/frozen_concat_linear.cpp:
```diff
@@ -0,0 +1,186 @@
+          auto candidate_weight = constant_as<Tensor>(candidate_node->namedInput("weight"));
+          auto candidate_weight_type = candidate_weight->dtype();
+          auto candidate_bias = constant_as<Tensor>(candidate_node->namedInput("bias"));
+          auto candidate_bias_type = candidate_bias->dtype();
```
We also need to be checking that the non-zero dims are equivalent (or all dims if we are not doing slicing logic)

Gamrix replied:
The optimization still works correctly if the non-zero dims are not equal, a thing that I intentionally programmed in. The unittest test this case.

eellison replied:
I don't think so ?
```
>>> inp = [torch.rand([1, 4]), torch.rand([0, 3])]
>>> torch.cat(inp)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: torch.cat(): Sizes of tensors must match except in dimension 0. Got 4 and 3 in dimension 1 (The offending index is 1)
>>> 
```

Gamrix replied:
```
inp = torch.rand([12, 5])
w = torch.rand([10, 5])
torch._C._nn.linear(inp, w)
w3 = torch.rand([10, 1])
torch._C._nn.linear(inp, w3)
```
throws the following error:
```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-5-59a7793913f5> in <module>
     13 torch._C._nn.linear(inp, w)
     14 w3 = torch.rand([10, 1])
---> 15 torch._C._nn.linear(inp, w3)

RuntimeError: mat1 and mat2 shapes cannot be multiplied (12x5 and 1x10)
```

Does that mean that this case would already be handled, or can we not assume that this is a checked condition? 
Might be prudent to add a check anyways.

eellison replied:
i think we should check that the tensors can be batched together - that the weights non-0 dim are equal. there might be other cases where the program would execute successfully  but not be easily batched, such as with broadcasting

eellison commented on torch/csrc/jit/passes/frozen_concat_linear.cpp:
```diff
@@ -0,0 +1,186 @@
+          long long next_loc = cur_loc + weight_tensor.size(0);
+          Value* next_val = graph_->insertConstant(next_loc);
+            
+          auto slice = graph_->create(
```
for any new Values we create, we would need to modify the AliasDb to maintain its correctness. I think for now I would either only accept linear layers which dont have modified inputs (this would not be a hard change), or regenerate the alias db on each transformation. We dont expect to do this vary frequently for any model so this might be fine but isn't the best....

if you imagine
```
x = linear(...)
y = linear(...)


x.add_(3)
```

after we combine x and y we would want the output of the first slice to still have its original aliasing properties. i think it would be simpler to just record which linear layer inputs / outputs are mutated, collect that in a list, and only optimize the layers that dont have mutated inputs. after we merge linear layers together we update the mutated list.

Gamrix replied:
Would it be fine if we processed the input tensor groups in reverse topological order (maybe that is too complicated)? Otherwise, I am happy to just go with the first option and keep it simple.

eellison replied:
Yea, i think reverse topological order would work as well

Gamrix replied:
[Note to self: Don't forget to add a test case where we have two interleaved optimizations needed so that I can test that this works correctly.]

eellison reviewed (commented):
Looks great!!

Gamrix committed:
Update on "Working on concatting multiple linear layers together"


Summary:
Linear layers using the same input tensor can be concatted together
as long as the weights and biases are compatible.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Gamrix committed:
Update on "Working on concatting multiple linear layers together"


Summary:
Linear layers using the same input tensor can be concatted together
as long as the weights and biases are compatible.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/frozen_concat_linear.cpp:
```diff
@@ -0,0 +1,255 @@
+      if (compatible_layers.size() == 1) {
+        continue; // No other layers to merge
+      }
+      mergeLinearLayers(base_node, compatible_layers);
```
if base node is in `compatible_layers` it makes the api a little confusing that we're passing it in twice..

eellison commented on torch/csrc/jit/passes/frozen_concat_linear.cpp:
```diff
@@ -0,0 +1,255 @@
+    }
+
+    // Processing for the block itself
+    std::unordered_map<Value*, std::list<Node*>> grouped_linear_layers;
```
So, the way that this would be a reverse topological iteration through the graph is if we did
```
block0:
     node1
     node2
           block1 
     node3
           block2
                  block3
```
block3->block2->node3->block1->node2->node1. However, what we are doing instead is block3->block2->block1->node3->node2->node1. I think we should be fine since we dont descend into block1 in alias analysis when we are processing node1/node2.... I think we are fine here since we are only looking at linear nodes in the same block and wouldnt look at any nodes in block1 while we are iterating block0, and we don't descend into blocks when trying the tryMoveTopologicallyValid as of now... just something to think about, we might want to add a test to make sure that if  alias analysis tryMoveTopologicallyValid did change we dont get an error. 


Something like :
```
def foo(x, cond: bool):
     b = c = torch.zeros([20, 20])
     bias = torch.zeros([20])
     out3 = torch.nn.functional.linear(x, b, bias)
     if cond:
          out1 = torch.nn.functional.linear(x, b, bias)
          out2 = torch.nn.functional.lienar(x, c, bias)
    else:
          assert False
    out4 = torch.nn.functional.linear(x, c, bias)
    return out1, out2, out3, out4
```
 
This way we check that trying to merge together out4 and out3 doesnt yield an error

eellison commented on test/jit/test_freezing.py:
```diff
@@ -2048,7 +2166,6 @@ def forward(self, x):
             scripted = torch.jit.freeze(torch.jit.script(mod))
             optimized = torch.jit.optimize_for_inference(scripted)
             inp = torch.rand([20, 20])
-            print(optimized.graph)
```
thx lol

eellison commented on test/jit/test_freezing.py:
```diff
@@ -1585,6 +1585,124 @@ def forward(self, x):
             test_conv_fusion(use_bias, nn.Conv2d, False, pytorch_op, False,
                              add_tensor=torch.rand(1).to(torch.int), expect_success=False)
 
+    # @unittest.skipIf(not torch._C.has_cuda, "Optimization currently only run for GPU")
```
commented out code

eellison commented on test/jit/test_freezing.py:
```diff
@@ -1585,6 +1585,124 @@ def forward(self, x):
+            mod_eager = ModMultLinear(w1_dim, w2_dim).eval()
+            script_mod = torch.jit.script(mod_eager)
+            op_graph = script_mod.graph
+            print(op_graph)
```
stray print

eellison commented on test/jit/test_freezing.py:
```diff
@@ -1585,6 +1585,124 @@ def forward(self, x):
+    # @unittest.skipIf(not torch._C.has_cuda, "Optimization currently only run for GPU")
+    def test_linear_concat_complex(self):
+        """
+            Testing that the interleaving of mutliple optimizations does not 
```
nit: `mutliple` -> `multiple`

eellison commented on test/jit/test_freezing.py:
```diff
@@ -1585,6 +1585,124 @@ def forward(self, x):
+        mod_eager = ModMultLinear(5, 5).eval()
+        script_mod = torch.jit.script(mod_eager)
+        op_graph = script_mod.graph
+        print(op_graph)
```
stray print

eellison commented on test/jit/test_freezing.py:
```diff
@@ -1585,6 +1585,124 @@ def forward(self, x):
+                res2 = torch._C._nn.linear(in_tensor2, self.w2, self.b2)
+                return res1, res2
+
+        mod_eager = ModMultLinear(5, 5).eval()
+        script_mod = torch.jit.script(mod_eager)
+        op_graph = script_mod.graph
+        print(op_graph)
+
+        FileCheck().check_count("aten::linear", 2, exactly=True).run(op_graph)
+        # successively no-ops with non-const inputs
+        self.run_pass("concat_frozen_linear", op_graph)
+        FileCheck().check_count("aten::linear", 2, exactly=True).run(op_graph)
+
+        script_mod = torch.jit.freeze(script_mod)
+        op_graph = script_mod.graph
+        self.run_pass("concat_frozen_linear", op_graph)
+        FileCheck().check_count("aten::linear", 2, exactly=True).run(op_graph)
+
+        test_val1 = torch.rand([50, 5])
+        test_val2 = torch.rand([50, 5])
+        self.assertEqual(mod_eager(test_val1, test_val2), script_mod(test_val1, test_val2))
```
I would factor out this block to take in a module, tuple of inputs, and number of linear nodes before/after calling the pass, since we are using this block 3 times.

eellison reviewed (commented):
Looks great!!! I have some small comments but I think we're just about good to land

Gamrix commented on torch/csrc/jit/passes/frozen_concat_linear.cpp:
```diff
@@ -0,0 +1,255 @@
+      Block* b,
+      std::unordered_map<Value*, std::list<Node*>>& grouped_linear_layers,
+      std::vector<Value*>& ordered_tensor_inputs) {
+    // We are using an ordered list so that one knows that
```
Using a list because it is linear time to remove from the middle of a list. I don't do any indexing into the data structure. (but I am not opposed to using a vector.

eellison replied:
yea, i comment on that elsewhere. i think removing and iterating through the list is harder/more complicated to follow and slower than the alternative

Gamrix commented on torch/csrc/jit/passes/frozen_concat_linear.cpp:
```diff
@@ -0,0 +1,255 @@
+        }
+
+        // Found a node that is eligible for combination
+        auto prev = it;
```
Because I didn't read the api for `erase` closely enough to realize it returns an iterator (oops)

Gamrix commented on torch/csrc/jit/passes/frozen_concat_linear.cpp:
```diff
@@ -0,0 +1,255 @@
+
+    // Now actually merge the candidate nodes
+
+    // Scope needed to make sure we free the WithInsertPoint guard
```
That doesn't work in this case, as `base_node` no longer exists at the end of the function when the destructor gets called to reset the insert_point. This causes a failed assertion.

Gamrix committed:
Update on "Concatting multiple linear layers with same input Tensor (different weight/bias)"


Summary:
Linear layers using the same input tensor can be concatted together
as long as the weights and biases are compatible.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Gamrix committed:
Update on "Concatting multiple linear layers with same input Tensor (different weight/bias)"


Summary:
Linear layers using the same input tensor can be concatted together
as long as the weights and biases are compatible.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Gamrix committed:
Update on "Concatting multiple linear layers with same input Tensor (different weight/bias)"


Summary:
Linear layers using the same input tensor can be concatted together
as long as the weights and biases are compatible.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/frozen_concat_linear.cpp:
```diff
@@ -0,0 +1,249 @@
+
+ public:
+  bool run(std::shared_ptr<Graph> graph) {
+    graph_ = graph;
```
nit: the conventions with other passes is to pass `ConcatLinearLayers(Graph)` and then have `run` not take in a graph. 

Currently, you allow passing in instantiating the class, passing in one `Graph` via run, and then passing in a second `Graph` into the same instantiated class. `graph_` will get reset, but not the other member fields, so you'd using the wrong `AliasDb` and `graph_modified` would be stale.

eellison commented on test/jit/test_freezing.py:
```diff
@@ -1585,6 +1584,135 @@ def forward(self, x):
+        test_val2 = torch.rand([50, 5])
+        self.check_linear_optimizations(mod_eager, 4, 3, (test_val1, test_val2, True))
+
+    def check_linear_optimizations(self, eager_mod, orig_linears, new_linears, test_vals):
```
nice!

eellison reviewed (approved):
🚢 🚢 🚢

Gamrix committed:
Update on "Concatting multiple linear layers with same input Tensor (different weight/bias)"


Summary:
Linear layers using the same input tensor can be concatted together
as long as the weights and biases are compatible.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Gamrix commented:
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31240642).

Gamrix committed:
Update on "Concatting multiple linear layers with same input Tensor (different weight/bias)"


Summary:
Linear layers using the same input tensor can be concatted together
as long as the weights and biases are compatible.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D31240642](https://our.internmc.facebook.com/intern/diff/D31240642)

[ghstack-poisoned]

Gamrix commented:
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31240642).

Gamrix committed:
Update on "Concatting multiple linear layers with same input Tensor (different weight/bias)"


Summary:
Linear layers using the same input tensor can be concatted together
as long as the weights and biases are compatible.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D31240642](https://our.internmc.facebook.com/intern/diff/D31240642)

[ghstack-poisoned]

Gamrix commented:
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31240642).

Gamrix committed:
Update on "Concatting multiple linear layers with same input Tensor (different weight/bias)"


Summary:
Linear layers using the same input tensor can be concatted together
as long as the weights and biases are compatible.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D31240642](https://our.internmc.facebook.com/intern/diff/D31240642)

[ghstack-poisoned]

Gamrix commented:
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31240642).

Gamrix committed:
Update on "Concatting multiple linear layers with same input Tensor (different weight/bias)"


Summary:
Linear layers using the same input tensor can be concatted together
as long as the weights and biases are compatible.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D31240642](https://our.internmc.facebook.com/intern/diff/D31240642)

[ghstack-poisoned]

Gamrix commented:
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31240642).
